In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-

# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
query = '''
    SELECT *
    FROM books
    '''

pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
query = '''
    SELECT *
    FROM authors
    '''

pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
query = '''
    SELECT *
    FROM ratings
    '''

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
query = '''
    SELECT *
    FROM reviews
    '''

pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [7]:
query = '''
    SELECT *
    FROM publishers
    '''

pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

**Анализ базы данных крупного сервиса для чтения книг по подписке.**

Наша компания выкупила крупный сервис для чтения книг по подписке в период наступления пандемии коронавируса. 

**Цель исследования**

Сформулировать ценностное предложение для вывода нового продукта на рынок электронных книг.
Для того чтобы наш бизнес развивался и процветал, нам необходимо всегда быть в тренде и предлагать актуальные продукты.


**Ход исследования**

•	Исследуем таблицы — выведите первые строки;

•	Сделаем по одному SQL-запросу для решения каждого задания;

•	Выведем результат каждого запроса в тетрадке;

•	Опишим выводы по каждой из решённых задач.


**Описание данных**

**Таблица books**

Содержит данные о книгах:

•	book_id — идентификатор книги;

•	author_id — идентификатор автора;

•	title — название книги;

•	num_pages — количество страниц;

•	publication_date — дата публикации книги;

•	publisher_id — идентификатор издателя.

**Таблица authors**

**Содержит данные об авторах:**

•	author_id — идентификатор автора;
•	author — имя автора.

**Таблица publishers**

**Содержит данные об издательствах:**

•	publisher_id — идентификатор издательства;

•	publisher — название издательства;

**Таблица ratings**

**Содержит данные о пользовательских оценках книг:**

•	rating_id — идентификатор оценки;

•	book_id — идентификатор книги;

•	username — имя пользователя, оставившего оценку;

•	rating — оценка книги.

**Таблица reviews**

**Содержит данные о пользовательских обзорах на книги:**

•	review_id — идентификатор обзора;

•	book_id — идентификатор книги;

•	username — имя пользователя, написавшего обзор;

•	text — текст обзора.


**Задания**

•	Посчитайте, сколько книг вышло после 1 января 2000 года;

•	Для каждой книги посчитайте количество обзоров и среднюю оценку;

•	Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

•	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

•	Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


### Исследование таблиц

#### Выведем первые строки

In [8]:
# первые 10 строк таблицы books
query = ''' 
    SELECT *
    FROM books
    LIMIT 10;
    ''' 

pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [9]:
# первые 10 строк таблицы books
query = ''' 
    SELECT *
    FROM authors
    LIMIT 10;
    ''' 

pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [10]:
# первые 10 строк таблицы ratings
query = ''' 
    SELECT *
    FROM ratings
    LIMIT 10;
    ''' 

pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [11]:
# первые 10 строк таблицы reviews
query = ''' 
    SELECT *
    FROM reviews
    LIMIT 10;
    ''' 

pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [12]:
# первые 10 строк таблицы publisher
query = ''' 
    SELECT *
    FROM publishers
    LIMIT 10;
    ''' 

pd.io.sql.read_sql(query, con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

#### Задание 1. Посчитайте, сколько книг вышло после 1 января 2000 года.

In [13]:
query = ''' 
    SELECT COUNT(DISTINCT book_id) AS cnt_books
    FROM books
    WHERE publication_date > '2000-01-01';
    ''' 

pd.io.sql.read_sql(query, con = engine)

cnt_books
0        819

После 1 января 2000 года вышло 819 книг. Пока нам не с чем сравнить . Знаем что сервис работает выпускает книги.Бизнес жив. 

#### Задание 2. Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [14]:
query = '''

SELECT b.book_id AS id,
       b.title AS bt,
       COUNT(DISTINCT r.username) AS cnt_rw,
       ROUND(AVG(rt.rating),2) AS AVG_rtn
FROM books AS b
FULL OUTER JOIN reviews AS r ON b.book_id = r.book_id
FULL OUTER JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY b.title, b.book_id
ORDER BY cnt_rw DESC;
 ''' 

pd.io.sql.read_sql(query, con = engine)

id                                                 bt  cnt_rw  avg_rtn
0    948                            Twilight (Twilight  #1)       7     3.66
1    963                                Water for Elephants       6     3.98
2    734                                   The Glass Castle       6     4.21
3    302  Harry Potter and the Prisoner of Azkaban (Harr...       6     4.41
4    695  The Curious Incident of the Dog in the Night-Time       6     4.08
..   ...                                                ...     ...      ...
995   83    Anne Rice's The Vampire Lestat: A Graphic Novel       0     3.67
996  808                            The Natural Way to Draw       0     3.00
997  672   The Cat in the Hat and Other Dr. Seuss Favorites       0     5.00
998  221                          Essential Tales and Poems       0     4.00
999  191  Disney's Beauty and the Beast (A Little Golden...       0     4.00

[1000 rows x 4 columns]

Мы получили список всех книг, у которых есть рейтинги и обзоры, и у которых нет ревью и рейтинга. Сумерки  7 обзоров и  средний рейтинг 3,66.

#### Задание 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [15]:
query = '''

SELECT p.publisher AS pub,
       COUNT(DISTINCT b.book_id) AS max_cnt_b
       
FROM publishers AS p
JOIN books AS b ON p.publisher_id = b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY max_cnt_b DESC
LIMIT 1;
 ''' 

pd.io.sql.read_sql(query, con = engine)

pub  max_cnt_b
0  Penguin Books         42

Издательство Penguin Books выпустило наибольшее число книнг толще 50 страниц - 42 книги

#### Задание 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [16]:
query = '''

SELECT a.author AS ath,
       AVG(r.rating) AS avg_r
       
FROM authors AS a
JOIN books AS b ON a.author_id = b.author_id
JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY a.author
HAVING COUNT(r.rating) > 50
ORDER BY avg_r DESC
LIMIT 1;
 ''' 

pd.io.sql.read_sql(query, con = engine)

ath     avg_r
0  J.K. Rowling/Mary GrandPré  4.288462

Джоан Кэ́тлин Роулинг автор которому присвоен самы высокий средний балл 4,288462 . Получила мировое признание после выпуска серии романов о Гарри Поттере.

#### Задание 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [17]:
query = '''

SELECT AVG(rw.rw_cnt) AS avg_cnt_rw       
FROM (
      SELECT username,
      COUNT(review_id) AS rw_cnt
      FROM reviews
      GROUP BY username
      HAVING username IN (
                          SELECT username
                          FROM ratings
                          GROUP BY username
                          HAVING COUNT(rating_id) > 50   
                        ) 
                 ) AS rw;
'''
pd.io.sql.read_sql(query, con = engine)

avg_cnt_rw
0   24.333333

В среднем 24,33 обзора публикуют пользователи которые поставил больше 50 оценок. Активные пользователи  - почти 50% обзоров от количества оценок.

#### Подведем итоги:

 - После 1 января 2000 года нашим книжным сервисом, было выпущено 819 книг.
 - Мы получили список всех книг, у которых есть рейтинги и обзоры, и у которых нет ревью и рейтинга. Сумерки  7 обзоров и  средний рейтинг 3,66.
 - Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц - 42 книги.
 - Джоан Кэ́тлин Роулинг автор которому присвоен самы высокий средний балл 4,288462 . Получила мировое признание после выпуска серии романов о Гарри Поттере.
 - В среднем 24,33 обзора публикуют пользователи которые поставил больше 50 оценок. Активные пользователи - почти 50% обзоров от количества оценок.